Document the triggering of a release step by step (#144)
gazebosim/docs: release_draft.md

#### Table of Contents

[Triggering a Release]()
  * [Prerequisites]()
     - [Software and Configurations]()
     - [Credentials]()

>   * [Preparing Ignition Software](

Reviewer: ```suggestion
  * [Preparing Ignition Software]()
```

>      - [Credentials](#credentials)

Reviewer 2: ```suggestion
     - [Access and Credentials](#access-and-credentials)
```

    and architecture) and uploads the .deb packages to
    `packages.osrfoundation.org` and [`osrf-distributions S3`](http://gazebosim.org/distributions).
 1. For macOS, the PR in `homebrew-simulation` waits for a comment from an
    Gz developer with the order `build bottle` that will trigger the job
    [`generic-release-homebrew_triggered_bottle_builder`](https://build.osrfoundation.org/job/generic-release-homebrew_bottle_builder/).
 1. `generic-release-homebrew_triggered_bottle_builder`will use the tarball with
    release sources from `osrf-distributions S3` to generate the binary bottles.
    They will be uploaded to `osrf-distributions S3`.

### Prerequisites

To perform a new release, a small number of configurations and credentials need to be made on the developer's system before triggering the release.
If a permanent operating system is used for releasing, these installation steps only need to be executed once.

#### Team and development checks

When creating a new release, there are some guidelines to follow before starting
the process:

 * Ask the team if there are any concerns about making the release.

>  * Check if there are changes to previous library versions that need to be forward-ported.

Reviewer 2: ```suggestion
 * Check if there are changes to previous library versions that need to be forward-ported. Forward-ports from the intended release library version are not important.
```

Author: Sorry I do not understand here what we mean by "release library".

 * See if there are open PRs against the release branch (release branch is the
   one with the name `ign/gz-fooX` where foo is the name of the Gz library and
   X the major version of the version bump planned) that could go into the new
   release.

#### Software and configurations

> Triggering the releasing process is only supported on Linux at this moment.

Reviewer 2: ```suggestion
> Triggering the releasing process is only supported on Linux at this moment.
```

The tool designed to facilitate the process of releasing software for all
platforms is called [`release.py`](https://github.com/gazebo-tooling/release-tools/blob/master/release.py), and is hosted at https://github.com/gazebo-tooling/release-tools.
Cloning the `release-tools` repository is required to perform a new release:

```
cd ~/
git clone https://github.com/gazebo-tooling/release-tools.git
```

The `release.py` tool will use `sc3md` software to interact with Amazon AWS
storage to host sources and binaries, so make sure `sc3md` is installed:

```
sudo apt-get install s3cmd
```

Some Debian tools require the following variables to be set:
```bash
export DEBEMAIL="<username>@openrobotics.org"
export DEBFULLNAME="<Your full name>"
```

> **Note:** the two above exported variables can be added to `.bashrc` to have
> them configured automatically on every run.

Reviewer 2: ```suggestion

> **Note:** the two above exported variables can be added to `.bashrc` to have
> them configured automatically on every run.
```

#### Credentials

There are some credentials needed to interact with the release
process:

 * S3 access to Open Robotics:
  ```
  s3cmd --configure
  ```
  If you don't have AWS credentials, please contact @j-rivero or @nuclearsandwich who will help set you up.

 * Release token: magic sequence of characters needed while running `release.py`
   to interact with `build.osrfoundation.org`. This should be given to Ignition releasers as a part of the AWS credentials set-up.

### Preparing Ignition Code

## Update code version

The first step to get a new release ready is to update the current code (upstream)

> version (view the [versioning](#versioning) section for more information). This

Author: fixed now.

bump could be in the major number (non-compatible changes), minor number (new
features), patch number (patches and bugfixes).

1. **Bumping major number** of the version implies some work to have the
[metadata](#metadata-for-releasing) updated correctly. There is a [dedicated

> document](releasing/bump_major.md) that you should go through before continuing to work through the steps in this

Author: The 'dedicated document' link should work once the PR is merged, AFAIK. Path seems correct to me:
```bash
docs on releasing_stepbystep [?] ❯ find . -name bump_major*
./releasing/bump_major.md
```

> document](releasing/bump_major.md) that you should go through before continuing to work through the steps in this

Reviewer 3: That's kinda weird, I don't see it in the branch :o 

https://github.com/gazebosim/docs/tree/releasing_stepbystep/releasing

> document](releasing/bump_major.md) that you should go through before continuing to work through the steps in this

Reviewer 3: Also I think it'd be nice if this section was separated or better emphasized, because I missed it the first time around and only realized because I knew I had to run the changelog generation in the main README

> document](releasing/bump_major.md) that you should go through before continuing to work through the steps in this

Author: Sorry for the missing file, I've uploaded it in this PR since it was only present locally.

> document](releasing/bump_major.md) that you should go through before continuing to work through the steps in this

Author: > Also I think it'd be nice if this section was separated or better emphasized, because I missed it the first time around and only realized because I knew I had to run the changelog generation in the main release-tools README

I've cleanup the index and added STEP in there, hopefully is easier now a4c7b71869cf422909ff3dea70be42f23ec2a88d

document.

   1. To update the upstream version a local checkout of the Ignition library is
   needed. A new branch is required to submit changes:

  ```bash
  # version X.Y.Z
  git checkout -b bump_to_X_Y_Z
  ```

  2. The current upstream version can be found in `CMakeLists.txt` file
  following the CMake `project declaration`:

  ```cmake
  # Ignition library named foo and version X.Y.Z
  project(ignition-fooX VERSION X.Y.Z)
  ```

  Stable releases can modify the X, Y or Z directly while prereleases will need to
  include the preX (X number starts with 1) suffix in the `ign_configure_project`:
  ```cmake
  # first prerelease of a serie, number 1
  ign_configure_project(VERSION_SUFFIX pre1)
  ```

  3. Together with bumping the version number **updating the Changelog and Migration
  documents** is highly recommended. `Changelog.md` file and `Migration.md` files
  are located at the top level of every Ignition library. Modify them as needed.

  ```bash
  git commit CMakeList.xt
  git commit Changelog.md
  git commit Migration.md

>   git push orign ...

Reviewer: ```suggestion
  git push origin ...
```

### Executing release.py

Make sure you are in the source code repository before running `release.py`. You should be on the branch to be released, after the pull request bumping the version has been merged (run `git status` to check the branch, and `git log` to check that the version bump pull request has been included). Running `release.py` from the source code repository will generate and upload some Git tags ("release tags") to the source code repository.

You will also need the token described in the [credentials section](#credentials).

#### dry-run simulation mode

The `release.py` tool supports a `--dry-run` flag that allows users to simulate releases
(nothing is modified) in order to ensure that the correct arguments are being used to trigger a particular release.
Ignition releasers should **always** call `release.py` with `--dry-run` first in order to ensure that proper commands are being used to trigger releases.

The script needs to be run from the repository with the source code (i.e., the repository where the Ignition library version bump pull request took place):

```bash
# Example of dry-run for ign-cmake2 bumped to 2.0.1
cd ign-cmake2
git checkout ign-cmake2

> ~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.1 <test_token> --dry-run

Reviewer: I had a hard time undertanding what is the `<test_token>`. Maybe we should add a note about this, how to get it, who can provide this, etc.

Author: 2a4f112

> ~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.1 dry-run-fake-token --dry-run

Reviewer 3: The release script is now in the [repo root](https://github.com/gazebo-tooling/release-tools/blob/master/release.py) :o

`~/release-tools/release.py`

Author: [7653d1c](https://github.com/gazebosim/docs/pull/144/commits/7653d1c1bf447a44348aa41a0c210e3666e190e4)

```

#### release.py for stable releases

```bash
# Ignition library named foo and major version X
cd ign-fooX
git checkout ign-fooX

# Example ign-cmake2 bumped to 2.0.1 with test_token credential
cd ign-cmake2
git checkout ign-cmake2
# please replace <test_token> with real release token (check crendentials section)
~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.1 <test_token>
```

#### release.py for prereleases or nightly

When releasing prereleases or nightly releases, there are some special flags to
be set. The `--upload-to-repo` argument is mandatory when running `release.py`, and
should be set to `prerelease` or `nightly`.

```bash
# Example ign-cmake2 bumped to prerelease 2.0.0~pre1 with test_token credential
cd ign-cmake2
git checkout ign-cmake2
# please replace <test_token> with real release token (check crendentials section)

> ~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.0~pre1 test_token --upload-to-repo prerelease

Reviewer 4: ```suggestion
~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.0~pre1 <test_token> --upload-to-repo prerelease
```

```
Nightly invocation is generally coded in the server. The version will be taken from

> the last changelog entry and [modified during building](releasing/versioning_pre_nightly).

Reviewer 4: this link is broken

No source code will be uploaded, but taken directly in the binary build from
`--nightly-src-branch`.

```bash
# Example ign-cmake2 nightly from main branch with test_token credential
cd ign-cmake2
git checkout ign-cmake2
# please replace <test_token> with real release token (check crendentials section)

> ~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.0~pre1 test_token --upload-to-repo nightly --nightly-src-branch main

Reviewer 4: ```suggestion
~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.0~pre1 <test_token> --upload-to-repo nightly --nightly-src-branch main
```

```
#### release.py for revision bumps

> Bumping the [revision number for binary packages](#versioning) is a special case of releasing

Reviewer 4: this link is broken

Author: should be fixed now

since the original tarball with the source code will remain the same. Once the
release repository is ready with the new release version, `release.py` needs the
`--only-bump-revision-linux` flag:

```bash
# Example ign-cmake2 bumped from 2.0.1-1 to 2.0.1-2 with test_token credential
cd ign-cmake2
git checkout ign-cmake2
# please replace <test_token> with real release token (check crendentials section)

> ~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.1 test_token --only-bump-revision-linux -release-version 2

Reviewer 4: ```suggestion
~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.1 <test_token> --only-bump-revision-linux -release-version 2
```

> ~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.1 test_token --only-bump-revision-linux -release-version 2

Author: sorry for that, changed to use a literal fake token [2a4f112](https://github.com/ignitionrobotics/docs/pull/144/commits/2a4f112355ec07afd9a74672ae8cef536f0c9e73)

> ```

Reviewer 4: after the `release.py` example sections, should we add some information about how to view the status of the builds that were triggered so that you can make sure everything succeeds? Currently, we have some information about this [here](https://github.com/ignition-tooling/release-tools/blob/a566144052ea70964f4b77cc34886395c50286f0/README.md) (point 10 at the end of the document - starts with "check that:")

Reviewer 2: Yeah I think a checklist would be best

Author: I added the previous info in 5061539f5c4e0435b78fd07af1b81ec077ceadce

When creating a new release there are some guidelines to follow before starting
the process:

 * Asking the team if there are any concerns about making the release
 * Check if there are changes to previous versions that need to be forward-ported.
 * See if there are open PRs against that release branch (release branch is the
   one with the name `ign-fooX` where foo is the name of the Ignition library and
   X the major version of the version bump planned) that could go into the new
   release.

#### Software and configurations

The releasing process is only supported in Linux at this moment.

The tool designed to facilitate the process of releasing software for all
platforms is called `release.py` and is hosted in the
[release-tools](https://github.com/ignition-tooling/release-tools/blob/master/release.py)
repository. Installing a copy of the repository is required to perform a new
release:

```
git clone https://github.com/ignition-tooling/release-tools.git
```

The `release.py` tool will use `sc3md` software to interact with Amazon AWS
storage to host sources and binaries.

```
sudo apt-get install s3cmd
```

Some Debian tools require the following variables to be set:
```bash

> export DEBEMAIL="<username>@openrobotics"

Reviewer 5: Should be a valid email address?

```suggestion
export DEBEMAIL="<username>@openrobotics.org"
```

export DEBFULLNAME="<Your full name>"
```
**Note:** the two above exported variables can be added to `.bashrc` to have
them configured automatically on every run.

#### Credentials

There is a small number of credentials needed to interact with the release
process:

 * S3 access to open robotics:
  ```
  s3cmd --configure
  ```

>   If you don't have AWS credentials please contact with TODO

Reviewer 5: @j-rivero we'll get some internal documentation on how to set up accounts for folks.
In the meantime are you okay being mentioned like this in public documentation or do you have a suggested alternative?

```suggestion
  If you don't have AWS credentials please contact @j-rivero or @nuclearsandwich who will help set you up.
```

Author: > @j-rivero we'll get some internal documentation on how to set up accounts for folks.
> In the meantime are you okay being mentioned like this in public documentation or do you have a suggested alternative?

sure no problem. We can start with this and see how it works, I don't expect too much activity with AWS credentials so anything simple sound good.to me.

build in the server. Different actions needs to happen:

 1. Generate a tarball with the Ignition library sources corresponding to the new
    version in the Ignition developer local system. Upload the tarball to S3 cloud storage.
 1. Request `build.osrfoundation.org` server to start the jobs for:
   1. Debian/Ubuntu: use `ign-fooX-debbuilder` job names
   1. Brew: entry job is `generic-release-homebrew_pull_request_updater`

The `release.py` script perform with all these actions.

### Executing release.py

`release.py` works from Ignition library code where the new version has been
set. Be sure to run it from a clone of a permanent branch (`ign-fooX` typically)
since some git tags will be generated and uploaded.

From the source code directory execute the `release.py`. You will need the token
described in the [credentials section](#credentials).

```bash
# Ignition library named foo and major version X
cd ign-fooX
git checkout ign-fooX

# Example ign-cmake2 bumped to 2.0.1 with test_token credential
cd ign-cmake2
git checkout ign-cmake2
~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.1 test_token
```

> #### dry-run simulation mode

Reviewer 4: This section should be mentioned **before** the actual `release.py` commands above, not after (that way, we make sure people check their `release.py` command for correctness before triggering a release).

> `release.py` tool supports a `--dry-run` flag that allows to simulate releases
> (nothing is modified) to help with safe checks done by the tool and providing
> the user with the final calls that will be done to the server.
> 
> The script needs to be run from the library repository:

Reviewer 4: ```suggestion
The `release.py` tool supports a `--dry-run` flag that allows users to simulate releases
(nothing is modified) in order to ensure that the correct arguments are being used to trigger a particular release.
Ignition releasers should **always** call `release.py` with `--dry-run` first in order to ensure that proper commands are being used to trigger releases.

The script needs to be run from the repository with the source code (i.e., the repository where the Ignition library version bump pull request took place):
```

```bash

> # Example of dry-run for ign-cmake2 bumped to 2.0.1
> cd ign-cmake2
> git checkout ign-cmake2
> ~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.1 test_token --dry-run

Reviewer 4: ```suggestion
# Example of dry-run for ign-cmake2 bumped to 2.0.1
cd ign-cmake2
git checkout bump_to_2_0_1
~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.1 test_token --dry-run
```

Author: We always release from supported branches (i.e: ign-cmake2, ign-transport8). The `bump_to_X_Y_Z` branch is used in the "Update code Version" section which detailed that a new PR should be open and merged (see "Update binary version").

```

#### release.py for prereleases or nightly

> When releasing prereleases or nightly releases there are some special flags to
> be set. `--upload-to-repo` argument is mandatory when running `release.py` and
> should indicate `prerelease` or `nightly`.

Reviewer 4: ```suggestion
When releasing prereleases or nightly releases, there are some special flags to
be set. The `--upload-to-repo` argument is mandatory when running `release.py`, and
should be set to `prerelease` or `nightly`.
```

```bash

> # Example ign-cmake2 bumped to prerelease 2.0.0~pre1 with test_token credential
> cd ign-cmake2
> git checkout ign-cmake2
> ~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.0~pre1 test_token --upload-to-repo prerelease

Reviewer 4: ```suggestion
# Example ign-cmake2 bumped to prerelease 2.0.0~pre1 with test_token credential
# (you should replace test_token with the actual token)
cd ign-cmake2
git checkout bump_to_2_0_1
~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.0~pre1 test_token --upload-to-repo prerelease
```

Author: same comment from before: we use `ign-cmake2` to make releases from it. Not sure if we can improve the document to make it clearer.

Reviewer 4: Sounds good :+1: should we at least make a note that `test_token` should be replaced with the actual release token?

Author: sure, 722326c

```

> Nightly invocation is generally coded in the server. Version will be taken from
> the last changelog entry and [modified during building](releasing/versioning_pre_nightly).
> No source code will be uploaded but taken directly in the binary build from

Reviewer 4: ```suggestion
Nightly invocation is generally coded in the server. The version will be taken from
the last changelog entry and [modified during building](releasing/versioning_pre_nightly).
No source code will be uploaded, but taken directly in the binary build from
```

`--nightly-src-branch`.

```bash
# Example ign-cmake2 nightly from main branch with test_token credential
cd ign-cmake2
git checkout ign-cmake2
~/release-tools/release-repo-scripts/release.py ign-cmake2 2.0.0~pre1 test_token --upload-to-repo nightly --nightly-src-branch main
```
#### release.py for revision bumps

> Bump the [revision number for binary packages](#versioning) is a special case of releasing

Reviewer 4: ```suggestion
Bumping the [revision number for binary packages](#versioning) is a special case of releasing
```

#### Table of Contents

[Triggering a Release]()
  * [Prerequisites]()
     - [Software and Configurations]()
     - [Credentials]()
  * [Preparing Ignition Software](
     - [Update version]()
     - [Update binary version]()
  * [Launch the release in the building server]()
    - [Executing release.py]()
      - [dry-run simulation mode]()
      - [release.py for prerelease or nightlies]()
      - [release.py for revision bumps]()

## Triggering a Release

### Understanding the releasing process

The following picture shows the interactions triggered when using the release
tool `release.py` explained in this guide:

![release.py interactions](releasing/images/releasepy_execution.svg)

Actions for releasing a new version of library `foo` with major version `X`:

>  1. `release.py` will generate a local tarball with the source code of the new

Reviewer 6: It would be interesting to add links to most of this stuff. The `release.py` script, the various jobs in build.osrfoundation.org...

Author: agree de2a3997b9d651cddf8f6fc9611aa38dcf9c3efd

  project(ignition-fooX VERSION X.Y.Z)
  ```

  Stable releases can modify the X, Y or Z directly while prereleases will need to
  include the preX (X number starts with 1) suffix in the `ign_configure_project`:
  ```cmake
  # first prerelease of a serie, number 1
  ign_configure_project(VERSION_SUFFIX pre1)
  ```

  3. Together with bumping the version number **updating the Changelog and Migration
  documents** is highly recommended. `Changelog.md` file and `Migration.md` files
  are located at the top level of every Ignition library. Modify them as needed.

  ```bash
  git commit CMakeList.xt
  git commit Changelog.md
  git commit Migration.md
  git push orign ...
  ```

  4. Open a pull request for reviewing ([example PR](https://github.com/ignitionrobotics/ign-physics/pull/132)).
  Including a link is recommended comparing the current release branch to the
  latest release. Releases are tagged in GitHub repositories with the scheme
  `ignition-fooX_X.Y.Z` where foo is the name of the Ignition library and X.Y.Z
  the code version. ([example of a branch comparison](https://github.com/ignitionrobotics/ign-gazebo/compare/ignition-gazebo3_3.5.0...ign-gazebo3)

>   5. Check that the CI job named `*-abichecker-*` is fine since its specially

Reviewer 6: Does the ABI checker check against the latest released tag or against the target release branch? I believe it's still the latter, right? See https://github.com/ignition-tooling/release-tools/issues/24#issuecomment-611519063.

If so, the ABI checker on the release PR will never fail, because that PR doesn't change C++ code...

Author: you are right, removed d8b482c4d83a394d0c82a1b654e9c0f2c60cae92
